crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -1140,4 +1339,206 @@ mod tests {
+        let spam_module = resolve_module(&db, spam_module_name.clone()).unwrap();
+
+        assert_eq!(
+            Some(foo_module.file()),
+            system_path_to_file(&db, site_packages.join("../x/y/src/foo.pyi"))
+        );
+        assert_eq!(
+            Some(a_module.file()),
+            system_path_to_file(&db, SystemPathBuf::from("/a.py"))
+        );
+        assert_eq!(
+            Some(b_module.file()),
+            system_path_to_file(&db, SystemPathBuf::from("/baz/b.py"))
+        );
+        assert_eq!(
+            Some(spam_module.file()),
+            system_path_to_file(&db, site_packages.join("spam/spam.py"))
+        );
+    }
+
+    #[test]
+    fn module_resolution_paths_cached_between_different_module_resolutions() {
+        const SRC: &[FileSpec] = &[("foo.py", "")];
+        const SITE_PACKAGES: &[FileSpec] = &[("bar.py", "")];
+
+        let TestCase {
+            mut db,
+            src,
+            site_packages,
+            ..
+        } = TestCaseBuilder::new()
+            .with_src_files(SRC)
+            .with_site_packages_files(SITE_PACKAGES)
+            .build();
+
+        let foo_module_name = ModuleName::new_static("foo").unwrap();
+        let bar_module_name = ModuleName::new_static("bar").unwrap();
+
+        let foo_module = resolve_module(&db, foo_module_name).unwrap();
+        assert_eq!(
+            Some(foo_module.file()),
+            system_path_to_file(&db, src.join("foo.py"))
+        );
+
+        db.clear_salsa_events();
+        let bar_module = resolve_module(&db, bar_module_name).unwrap();
+        assert_eq!(
+            Some(bar_module.file()),
+            system_path_to_file(&db, site_packages.join("bar.py"))
+        );
+        let events = db.take_salsa_events();
+        assert_function_query_was_not_run::<module_search_paths, _, _>(
+            &db,
+            |res| &res.function,
+            &(),
+            &events,
+        );
+    }
+
+    #[test]
+    fn deleting_pth_file_on_which_module_resolution_depends_invalidates_cache() {
+        const SITE_PACKAGES: &[FileSpec] = &[("_foo.pth", "/x/src")];
+        const X_DIRECTORY: &[FileSpec] = &[("/x/src/foo.py", "")];
+
+        let TestCase {
+            mut db,
+            site_packages,
+            ..
+        } = TestCaseBuilder::new()
+            .with_site_packages_files(SITE_PACKAGES)
+            .with_other_files(X_DIRECTORY)
+            .build();
+
+        let foo_module_name = ModuleName::new_static("foo").unwrap();
+        let foo_module = resolve_module(&db, foo_module_name.clone()).unwrap();
+        assert_eq!(
+            Some(foo_module.file()),
+            system_path_to_file(&db, SystemPathBuf::from("/x/src/foo.py"))
+        );
+
+        let pth_file = site_packages.join("_foo.pth");
+        db.memory_file_system().remove_file(&pth_file).unwrap();
+        File::touch_path(&mut db, &pth_file);
+        assert_eq!(resolve_module(&db, foo_module_name.clone()), None);
+    }
```

Comment:
This test is failing, and I'm not sure why. To read the contents of the `pth` files, I'm using `system_path_to_file` in `PthFileIterator::next()`, and `source_text()` in `PthFile::new()`. IIUC, I think that should be sufficient to indicate to Salsa that the `module_search_paths()` query depends on the existence and contents of the `_foo.pth` file, and that therefore deleting the `_foo.pth` file should invalidate the previous query.

Reply:
I took a quick look at the salsa events and this seems correct. 

```
[crates/red_knot_module_resolver/src/resolver.rs:1555:9] db.take_salsa_events().debug(&db) = [
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: WillCheckCancellation,
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: WillCheckCancellation,
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: WillCheckCancellation,
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: DidValidateMemoizedValue {
            database_key: source_text(2),
        },
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: DidValidateMemoizedValue {
            database_key: parse_typeshed_versions(2),
        },
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: WillCheckCancellation,
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: WillExecute {
            database_key: dynamic_module_resolution_paths(0),
        },
    },
    Event {
        runtime_id: RuntimeId {
            counter: 0,
        },
        kind: DidValidateMemoizedValue {
            database_key: resolve_module_query(0),
        },
    },
]
```

It re-runs the `dynamic_module_resolution_paths` query because the dynamic module resolution paths changed (you deleted a `phf` file). It then re-runs the `resolve_module_query` but evaluates to the same cached result. Not sure why this is the case.

Reply:
It seems that adding `db.report_untracked_read()` to the function body, like you suggested elsehwere, makes the test start passing. But then if I start using `db.system().read_to_string()` in the query to get the contents of the `.pth` file rather than `source_text()` (which you said _should_ be okay if I added the `report_untracked_read()` call), it starts failing again.